src/commonMain/kotlin/com/charleskorn/kaml/YamlNodeSerializer.kt

```diff
@@ -0,0 +1,172 @@
+
+*/
+
+@file:OptIn(ExperimentalSerializationApi::class, InternalSerializationApi::class)
+
+package com.charleskorn.kaml
+
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.InternalSerializationApi
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.builtins.ListSerializer
+import kotlinx.serialization.builtins.MapSerializer
+import kotlinx.serialization.builtins.serializer
+import kotlinx.serialization.descriptors.PolymorphicKind
+import kotlinx.serialization.descriptors.PrimitiveKind
+import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.SerialKind
+import kotlinx.serialization.descriptors.buildSerialDescriptor
+import kotlinx.serialization.descriptors.nullable
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.encoding.encodeStructure
+
+internal object YamlNodeSerializer : KSerializer<YamlNode> {
+    override val descriptor: SerialDescriptor =
+        buildSerialDescriptor("com.charleskorn.kaml.YamlNode", PolymorphicKind.SEALED) {
+            annotations += YamlContentPolymorphicSerializer.Marker()
+        }.nullable
+
+    override fun serialize(encoder: Encoder, value: YamlNode) {
+        encoder.asYamlOutput()
+        when (value) {
+            is YamlList -> encoder.encodeSerializableValue(YamlListSerializer, value)
+            is YamlMap -> encoder.encodeSerializableValue(YamlMapSerializer, value)
+            is YamlNull -> encoder.encodeSerializableValue(YamlNullSerializer, value)
+            is YamlScalar -> encoder.encodeSerializableValue(YamlScalarSerializer, value)
+            is YamlTaggedNode -> encoder.encodeSerializableValue(YamlTaggedNodeSerializer, value)
+        }
+    }
+
+    override fun deserialize(decoder: Decoder): YamlNode {
+        val input = decoder.asYamlInput<YamlInput>()
+        return if (input is YamlPolymorphicInput) YamlTaggedNode(input.typeName, input.node) else input.node
+    }
+}
+
+internal object YamlScalarSerializer : KSerializer<YamlScalar> {
+    override val descriptor: SerialDescriptor =
+        PrimitiveSerialDescriptor("com.charleskorn.kaml.YamlScalar", PrimitiveKind.STRING)
+
+    override fun serialize(encoder: Encoder, value: YamlScalar) {
+        encoder.asYamlOutput()
+        try {
+            return encoder.encodeBoolean(value.toBoolean())
+        } catch (_: SerializationException) {
+        }
+        try {
+            return encoder.encodeLong(value.toLong())
+        } catch (_: SerializationException) {
+        }
+        try {
+            return encoder.encodeDouble(value.toDouble())
+        } catch (_: SerializationException) {
+        }
+        value.content.singleOrNull()?.also { return encoder.encodeChar(it) }
+        encoder.encodeString(value.content)
```

Comment:
Is there a test for this special handling of characters?

Reply:
No, I just wanted to use the short version of `encodeChar`, to not have the overhead from `encodeString`. There are also no extra tests for `encodeBoolean`, `encodeLong`, `encodeDouble`. I thought, these are properly tested in the tests for `YamlOutput`.
But I can of course write extra tests, if you think, that's needed.

Reply:
I think it'd be valuable to have a test that shows the expected behaviour when a value looks like it is a boolean or a number etc.

Reply:
I added some tests for encoding the primitives via the new serializer.